Log Replica Changes to Rangelog table

mrtracy opened the pull request:
"ChangeReplicas" events are now logged to the system.rangelog table. This will
allow tracking of rebalances and other replication changes.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4085)

<!-- Reviewable:end -->

Commit:
Log Replica Changes to Rangelog table

"ChangeReplicas" events are now logged to the system.rangelog table. This will
allow tracking of rebalances and other replication changes.

tamird commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[storage/log.go, line 155 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YC9JJvzmGsyCC1-5W:-K9YC9JJvzmGsyCC1-5X:192171377) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log.go#L155)):*
nit: switch might be nicer here.

---

*[storage/log.go, line 185 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YCPWRcOwCLyvgfsH3:-K9YCPWSUADdcpo3UEwu:-2019193392) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log.go#L185)):*
you should put this method in storage/helpers_test so that it is only linked in in tests.

---

*[storage/log.go, line 195 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YBohDWOTB9edSWcGk:-K9YBohDWOTB9edSWcGl:-1693987330) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log.go#L195)):*
nit: I think you should operate on the roach.Timestamp here, since `roachpb.ZeroTimestamp` is already a thing.

This function could do with a better name.

---

*[storage/log_test.go, line 128 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YCbq1DafX9Sh67X7j:-K9YCbq1DafX9Sh67X7k:-1085152496) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L128)):*
either remove or complete

---

*[storage/log_test.go, line 177 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YCxGnkJHB_6c2Lntf:-K9YCxGnkJHB_6c2Lntg:1432028433) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L177)):*
are placeholderes not work again?

---

*[storage/log_test.go, line 196 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDKVs9n4y0_76S3R3:-K9YDKVs9n4y0_76S3R4:-1053298648) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L196)):*
not split

---

*[storage/log_test.go, line 207 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDIwNvcR0iKRVoTTw:-K9YDIwNvcR0iKRVoTTx:1707480395) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L207)):*
not splits

---

*[storage/log_test.go, line 210 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDJgLB9ggcMMMMYBv:-K9YDJgLB9ggcMMMMYBw:1707480395) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L210)):*
not splits

---

*[storage/log_test.go, line 213 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDP1EbWzWDG4pW_Z0:-K9YDP1F8dle5xOeHOnH:-724030747) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L213)):*
nit: this should be above any `continue`s, or you'll get spurious errors.

---

*[storage/log_test.go, line 219 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDTvKiThDR-gPTtrX:-K9YDTvKiThDR-gPTtrY:-1115167666) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L219)):*
why fatal suddenly?

---

*[storage/log_test.go, line 236 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDYVzGsnqFkiHt7zX:-K9YDYVzGsnqFkiHt7zY:-18990801) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L236)):*
not add

---

*[storage/log_test.go, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDZDcmiNZo425dSDS:-K9YDZDcmiNZo425dSDT:-1053298648) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L240)):*
not split

---

*[storage/log_test.go, line 247 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YD_2UrgJf1_VboVhR:-K9YD_2UrgJf1_VboVhS:-18990801) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L247)):*
not add

---

*[storage/log_test.go, line 251 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YD_hP4hB_GTxy7zaO:-K9YD_hP4hB_GTxy7zaP:-1053298648) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L251)):*
not split

---

*[storage/log_test.go, line 254 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDa_XKo99hOfW0231:-K9YDa_XKo99hOfW0232:-605495130) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L254)):*
not split

does `%v` actually present a useful form here?

---

*[storage/log_test.go, line 257 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDpwOPh_E28qcfDHX:-K9YDpwOPh_E28qcfDHY:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L257)):*
ditto

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/log_test.go, line 177 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YCxGnkJHB_6c2Lntf:-K9YHXD3-jeYYmCkBRR6:1438010344) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L177)):*
I think @mjibson is already looking (from one of Pete's PRs).

---

*[storage/log_test.go, line 183 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YHb8p2GGWJ9thfymd:-K9YHb8p2GGWJ9thfyme:789161313) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L183)):*
nit: `var count int`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 155 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YC9JJvzmGsyCC1-5W:-K9YkYwHWHhYrcml7lEC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log.go#L155)):*
Done.

---

*[storage/log.go, line 185 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YCPWRcOwCLyvgfsH3:-K9YgFt-m7PLpQLU4yi-:-895990265) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log.go#L185)):*
It's used by tests that are not in the storage package, it has to go in the prod file. 

---

*[storage/log.go, line 195 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YBohDWOTB9edSWcGk:-K9YldpKRp7WZ2YKmfwD:123776263) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log.go#L195)):*
Fixed, changed name to `selectEventTimestamp`

---

*[storage/log_test.go, line 128 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YCbq1DafX9Sh67X7j:-K9Yljz36pDlvD3I1AOk:1675637715) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L128)):*
removed.

---

*[storage/log_test.go, line 177 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YCxGnkJHB_6c2Lntf:-K9YgSY29JkKzT83lFzU:-349463295) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L177)):*
Yes, I meant to replace this comment with another TODO.

---

*[storage/log_test.go, line 183 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YHb8p2GGWJ9thfymd:-K9YmKb6Ck3bfpBcEzYb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L183)):*
Done.

---

*[storage/log_test.go, line 196 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDKVs9n4y0_76S3R3:-K9YmdUCT1UFUSTHRBQV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L196)):*
Done.

---

*[storage/log_test.go, line 207 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDIwNvcR0iKRVoTTw:-K9Yme-6-KAPKZsqp1IL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L207)):*
Done.

---

*[storage/log_test.go, line 210 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDJgLB9ggcMMMMYBv:-K9YmePRKqSIjfy-kREM:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L210)):*
Done.

---

*[storage/log_test.go, line 213 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDP1EbWzWDG4pW_Z0:-K9YnGdi8lCTVc6vHpAY:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L213)):*
Done.

---

*[storage/log_test.go, line 219 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDTvKiThDR-gPTtrX:-K9YnBj4HCKVSCRftoLp:-2121697940) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L219)):*
I guess it does work as an Errorf, switched.

---

*[storage/log_test.go, line 236 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDYVzGsnqFkiHt7zX:-K9YmfFLm1zh7vKexaSV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L236)):*
Done.

---

*[storage/log_test.go, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDZDcmiNZo425dSDS:-K9YmfVz4Y9ehDS-zspd:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L240)):*
Done.

---

*[storage/log_test.go, line 247 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YD_2UrgJf1_VboVhR:-K9YmfnU8tQKlVOuj4Fw:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L247)):*
Done.

---

*[storage/log_test.go, line 251 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YD_hP4hB_GTxy7zaO:-K9YmgIHKs9n6j-JxBHO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L251)):*
Done.

---

*[storage/log_test.go, line 254 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDa_XKo99hOfW0231:-K9YmlJENlG2XSt9X4-5:-364488963) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L254)):*
%v presents as '{%d %d %d}' with the rangeid, nodeid and replicaid

---

*[storage/log_test.go, line 257 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDpwOPh_E28qcfDHX:-K9YnI6ngQhNhcT77iOk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L257)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 185 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YCPWRcOwCLyvgfsH3:-K9YxKDoSvVHWwGnVRGb:-1963312612) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log.go#L185)):*
You mean because those tests are in storage_test? See the comment at the top of storage/helpers_test.go.

---

*[storage/log_test.go, line 254 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDa_XKo99hOfW0231:-K9YxvM3sSwseXySTS3e:-1519283120) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L254)):*
that seems just about impossible to decipher. why not use the stringer?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 2 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/log.go, line 185 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YCPWRcOwCLyvgfsH3:-K9Z1BXMn3s352WBoDZw:-1597764173) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log.go#L185)):*
Didn't know you could do that. Moved.

---

*[storage/log_test.go, line 254 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-K9YDa_XKo99hOfW0231:-K9Z0YmChGEeVVDj-RZJ:1505156304) ([raw file](https://github.com/cockroachdb/cockroach/blob/110d1e4d36ded7dc68ae5425526430478625aef9/storage/log_test.go#L254)):*
That's an interesting standard for "impossible". Nevertheless, changed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Assuming this is LGTM to somebody?

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 3 of 3 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4085#-:-K9cycaEBniypGwg49sE:980077423)*
<!-- Sent from Reviewable.io -->
